pkg/ddl/ddl.go

```diff
@@ -1285,6 +1407,99 @@ func (d *ddl) SwitchMDL(enable bool) error {
 	return nil
 }
 
+// SwitchDDLVersion switch ddl version.
+func (d *ddl) SwitchDDLVersion(version int64) error {
+	oldVersion := variable.DDLVersion.Load()
+	if oldVersion == version {
+		return nil
+	}
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*30)
+	defer cancel()
+
+	// Check if there is any DDL running.
+	// This check can not cover every corner cases, so users need to guarantee that there is no DDL running by themselves.
+	sessCtx, err := d.sessPool.Get()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer d.sessPool.Put(sessCtx)
+	se := sess.NewSession(sessCtx)
+	rows, err := se.Execute(ctx, "select 1 from mysql.tidb_ddl_job", "check job")
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if len(rows) != 0 {
+		return errors.New("please wait for all jobs done")
+	}
+
+	if err := d.migration2DDLVersion(version); err != nil {
+		return errors.Trace(err)
+	}
+
+	variable.DDLVersion.Store(version)
+	logutil.BgLogger().Info("switch ddl version", zap.String("category", "ddl"), zap.Int64("version", version))
+	return nil
+}
+
+// migration2DDLV1 migration ddl version to v1.
+func (*ddl) migration2DDLV1(m *meta.Meta) error {
+	// clear all table names when we switch to v1.
+	return errors.Trace(m.ClearAllTableNames())
+}
+
+// migration2DDLV2 migration ddl version to v2.
+func (*ddl) migration2DDLV2(m *meta.Meta) error {
+	if err := m.ClearAllTableNames(); err != nil {
+		return errors.Trace(err)
+	}
+
+	dbs, err := m.ListDatabases()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	for _, dbInfo := range dbs {
+		tables, err := m.ListTables(dbInfo.ID)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		for _, tableInfo := range tables {
+			if err := m.CreateTableName(dbInfo.Name.String(), tableInfo.Name.String(), tableInfo.ID); err != nil {
+				return errors.Trace(err)
+			}
+		}
+	}
+
+	return err
+}
+
+func (d *ddl) migration2DDLVersion(ver int64) error {
+	return kv.RunInNewTxn(kv.WithInternalSourceType(d.ctx, kv.InternalTxnDDL), d.store, true, func(ctx context.Context, txn kv.Transaction) error {
+		m := meta.NewMeta(txn)
+		v, err := m.GetDDLVersion()
```

Comment:
Why do we need this meta key?

Reply:
In order to achieve idempotence.

Reply:
This will also prevent the index from being created multiple times.

Reply:
I suggest renaming it to `NameMapInitialized` instead.